src/librustc_mir/borrow_check/borrow_set.rs

```diff
@@ -97,7 +100,24 @@ impl<'tcx> fmt::Display for BorrowData<'tcx> {
 }
 
 impl<'tcx> BorrowSet<'tcx> {
-    pub fn build(tcx: TyCtxt<'_, '_, 'tcx>, mir: &Mir<'tcx>) -> Self {
+    pub fn build(
+        tcx: TyCtxt<'_, '_, 'tcx>,
+        mir: &Mir<'tcx>,
+        locals_are_invalidated_at_exit: bool,
+        ) -> Self {
+        struct HasStorageDead(BitArray<Local>);
+
+        impl<'tcx> Visitor<'tcx> for HasStorageDead {
+            fn visit_local(&mut self, local: &Local, ctx: PlaceContext<'tcx>, _: Location) {
+                if ctx == PlaceContext::StorageDead {
+                    self.0.insert(*local);
+                }
+            }
+        }
+
+        let mut has_storage_dead = HasStorageDead(BitArray::new(mir.local_decls.len()));
+        has_storage_dead.visit_mir(mir);
```

Comment:
In a similar prototype that I was working on, I combined this bit-array computation into a new enum for `locals_are_invalidated_at_exit`. (That is, I tried to ensure that we only did this work when `locals_are_invalidated_at_exit` is false here.) Just a stray thought on something you might want to consider incorporating.

Reply:
agree, I will do this join as a next step, but may be the bit-array can be used somewhere else? or just make it an Option but I cannot find a nice name for it

Reply:
Well, if I understand correctly, you can only *have* the bit-array when `locals_are_invalidated_at_exit` is false in the first place. So even if the bit-array can be used somewhere else, you should be able to extract it, assuming you've changed the type like so: `locals_are_invalidated_at_exit: LocalInvalidationState`, where
```rust
 enum InvalidationState {
    LocalsAreInvalidated(BitArray<Local>),
    LocalsNotInvalidated,
}
```